examples/image_captioning/model.py

```diff
@@ -0,0 +1,267 @@
+
+        Compute the softmax cross-entropy captioning loss.
+        """
+        self.reset(img_feats)
+
+        loss = 0
+        size = 0
+        caption_length = captions.shape[1]
+        for i in range(caption_length - 1):
+            # Compute the loss based on the prediction of the next token in the
+            # sequence
+            x = Variable(self.xp.asarray(captions[:, i]))
+            t = Variable(self.xp.asarray(captions[:, i + 1]))
+            if (t.data == self.embed_word.ignore_label).all():
+                # The next tokens in the sequence should not be accounted for
+                # in the loss, e.g. the targets are paddings that were used to
+                # align the lengths of all captions within the same batch
+                break
+            y = self.step(x)
+            loss += F.softmax_cross_entropy(
+                y, t, ignore_label=self.embed_word.ignore_label)
+            size += 1
+        return loss / max(size, 1)
+
+    def predict(self, img_feats, bos, eos, max_caption_length):
+        """Batch of image features to captions."""
+        self.reset(img_feats)
+
+        captions = self.xp.full((img_feats.shape[0], 1), bos, dtype=np.int32)
+        for _ in range(max_caption_length):
+            x = Variable(captions[:, -1])  # Previous word token as input
+            y = self.step(x)
+            pred = y.data.argmax(axis=1).astype(np.int32)
+            captions = self.xp.hstack((captions, pred[:, None]))
+            if (pred == eos).all():
+                break
+        return captions
+
+    def reset(self, img_feats):
+        """Batch of image features to hidden representations.
+
+        Also, reset and then update the internal state of the LSTM.
+        """
+        self.lstm.reset_state()
+        h = self.embed_img(img_feats)
+        h = self.lstm(F.dropout(h, ratio=self.dropout_ratio))
+        return h
+
+    def step(self, x):
+        """Batch of word tokens to word tokens.
+
+        Predict the next set of tokens given previous tokens.
+        """
+        h = self.embed_word(x)
+        h = self.lstm(F.dropout(h, ratio=self.dropout_ratio))
+        h = self.out_word(F.dropout(h, ratio=self.dropout_ratio))
+        return h
+
+
+class NStepLSTMLanguageModel(chainer.Chain):
+    """Recurrent NStepLSTM language model.
+
+    Generate captions given features extracted from images.
+    """
+
+    def __init__(self, vocab_size, hidden_size, dropout_ratio, ignore_label):
+        super(NStepLSTMLanguageModel, self).__init__()
+        with self.init_scope():
+            self.embed_word = L.EmbedID(
+                vocab_size,
+                hidden_size,
+                initialW=initializers.Normal(1.0),
+                ignore_label=ignore_label
+            )
+            self.embed_img = L.Linear(
+                hidden_size,
+                initialW=initializers.Normal(0.01),
+                initial_bias=initializers.Zero()
```

Comment:
As well as I mentioned above, this line is not required.

Reply:
Removed as agreed.